### PR TITLE
Decrease presence of joda time

### PR DESCRIPTION
**Pull request by ssheikin**
## Description

Part of https://github.com/trinodb/trino/issues/14850

Get rid of joda time in TestTrinoDriver

Rewrite `convertUTCToLocal`, `convertLocalToUTC` methods with java.time
in modules:
`trino-hive-formats`
`trino-orc`
`trino-parquet`
`trino-plugin-toolkit`

Above changes are highly coupled, so are present in the same commit.

Those changes effectively decreases maven scope/eliminates joda-time
dependency presence in
`trino-delta-lake`
`trino-hudi`
`trino-parquet`
`trino-plugin-toolkit`
`trino-raptor-legacy`

## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

**Comment by martint** [comment]
Did you measure the impact in performance? The reason we originally didn’t remove joda everywhere is that the Java datetime APIs require some usage patterns that can be inefficient.

File: lib/trino-parquet/src/main/java/io/trino/parquet/predicate/TupleDomainParquetPredicate.java [hunk 1]
```diff
@@ -310,8 +311,8 @@ public static Domain getDomain(
             return getDomain(
                     column,
                     type,
-                    ImmutableList.of(statistics.genericGetMin()),
-                    ImmutableList.of(statistics.genericGetMax()),
+                    List.of(statistics.genericGetMin()),
```
Comment (raunaqmorarka) [inline lib/trino-parquet/src/main/java/io/trino/parquet/predicate/TupleDomainParquetPredicate.java:314]: unrelated change

File: lib/trino-parquet/src/main/java/io/trino/parquet/reader/decoders/ValueDecoders.java [hunk 2]
```diff
@@ -1255,4 +1258,13 @@ private IllegalArgumentException wrongEncoding(ParquetEncoding encoding)
     {
         return new IllegalArgumentException("Wrong encoding " + encoding + " for column " + field.getDescriptor());
     }
+
+    private static long convertUTCToLocal(ZoneId zoneId, long millis)
```
[the remark this answers is not in the document]
Comment (ssheikin) [reply lib/trino-parquet/src/main/java/io/trino/parquet/reader/decoders/ValueDecoders.java:1262]: Thank you @raunaqmorarka for pointing to benchmark! I've run it as you advised and here are the results:
```
DateTimeZone.UTC
Benchmark                           Mode  Cnt    Score   Error  Units
BenchmarkFixed12ColumnReader.read  thrpt   20  174.050 ± 1.355  ops/s

DateTimeZone.forID("Europe/Berlin")
Benchmark                           Mode  Cnt  Score   Error  Units
BenchmarkFixed12ColumnReader.read  thrpt   20  3.880 ± 0.132  ops/s

ZoneOffset.UTC
Benchmark                           Mode  Cnt    Score   Error  Units
BenchmarkFixed12ColumnReader.read  thrpt   20  173.663 ± 1.520  ops/s

ZoneId.of("Europe/Berlin")
Benchmark                           Mode  Cnt  Score   Error  Units
BenchmarkFixed12ColumnReader.read  thrpt   20  9.984 ± 0.144  ops/s
```
I see that `ZoneOffset.UTC` and `DateTimeZone.UTC` are comparable, but `ZoneId.of("Europe/Berlin")` is 2.5 times **faster** than `DateTimeZone.forID("Europe/Berlin")`.

File: lib/trino-hive-formats/src/main/java/io/trino/hive/formats/encodings/binary/BinaryColumnEncodingFactory.java [hunk 3]
```diff
@@ -23,8 +23,8 @@
 import io.trino.spi.type.TimestampType;
 import io.trino.spi.type.Type;
 import io.trino.spi.type.VarcharType;
-import org.joda.time.DateTimeZone;
```
Comment (ssheikin) [inline lib/trino-hive-formats/src/main/java/io/trino/hive/formats/encodings/binary/BinaryColumnEncodingFactory.java:26]: https://github.com/trinodb/trino/pull/18632#issuecomment-1673332891
>Did you measure the impact in performance? The reason we originally didn’t remove joda everywhere is that the Java datetime APIs require some usage patterns that can be inefficient.

@martint I've run benchmarks which @raunaqmorarka adviced.
https://github.com/trinodb/trino/pull/18632#discussion_r1291163921
So far I see performance improvement with this change.
I'd be glad to reuse knowledge gathered during previous attempts. 
Please point me where previous attempts took place. 
Could you please elaborate what are those inefficient usage patterns you are mentioning?

File: lib/trino-parquet/src/main/java/io/trino/parquet/reader/decoders/ValueDecoders.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (huberty89) [reply lib/trino-parquet/src/main/java/io/trino/parquet/reader/decoders/ValueDecoders.java:1262]: Where do you run this benchmark? If on your local machine does hardware where you running the benchmark does make a difference?

File: lib/trino-hive-formats/src/main/java/io/trino/hive/formats/encodings/binary/BinaryColumnEncodingFactory.java (hunk 3 above)
Comment (ssheikin) [reply lib/trino-hive-formats/src/main/java/io/trino/hive/formats/encodings/binary/BinaryColumnEncodingFactory.java:26]: @martint I see, you have concerns about additional objects creation and how JIT optimizes execution flows.
From the first glance joda has more math and less of objects creation (however still has).
Running tests provided by @raunaqmorarka show it's not a problem. 
Also those tests prove java.time is much faster than joda.

I'm more than happy to run tests and be sure that my changes do not bring regression, especially in performance.
However, I'm new to performance testing in trino. 
How is it usually done? Where I can find documentation about it? When/by whom usually those test run? 
Where I can find performance tests which justify joda presence?
What is the acceptance criteria for this PR?

Also I see that this PR may be more controversial that I anticipated and require more opinions to gather.
Could you please help me identify involved contributors/maintainers who has worked on the initial partial migration from joda to java.time or who can be interested in this topic? Should I move the discussion to the slack according to https://trino.io/development/process.html ?

I'm eagerly waiting for your answer as I really want to have this PR merged or to have a prove that it does not worth.

File: lib/trino-parquet/src/main/java/io/trino/parquet/reader/decoders/ValueDecoders.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (ssheikin) [reply lib/trino-parquet/src/main/java/io/trino/parquet/reader/decoders/ValueDecoders.java:1262]: Yes, on my local machine.

@huberty89 @raunaqmorarka @martint on what machines I should run performance tests for trino? How can I do that properly? What kind of machines we have to run tests for trino?

**Comment by mosabua** [comment]
@ssheikin are you still continuing this work?

**Comment by ssheikin** [comment]
Thank you @findepi for jumping in and describing the situation around the joda time in trino. 
Long long time ago in a Trino far, far away an effort to eliminate joda presence took place, but as @martint pointed, some extensive testing showed inefficiencies of java 8 implementation in some cases.
There are no testing tools to reuse left from those efforts and it requires investment in scrapping valuable ancient knowledge.
Happy to get back to this topic once there is enough sponsorship for it.
